pkg/hs/urbit-king/lib/Urbit/Vere/Term/Demux.hs

```diff
@@ -77,16 +88,45 @@ dGive Demux{..} evs = do
     If there are no attached clients, this will not return until one
     is attached.
 -}
-dTake :: Demux -> STM (Maybe Belt)
+dTake :: Demux -> STM (Maybe ClientTake)
 dTake Demux{..} = do
     conns <- readTVar dConns
-    waitForBelt conns >>= \case
-        (_, Just b ) -> pure (Just b)
-        (k, Nothing) -> do writeTVar dConns (ksDelete k conns)
-                           pure Nothing
+    waitForTake conns >>= \case
+        (_, Just (ClientTakeBelt b)) -> pure (Just (ClientTakeBelt b))
+
+        (k, Just (ClientTakeSize s)) -> do
+          newSizeTree <- modifyAndReturnTVar dSizes (insertMap k s)
+          maybeUpdateTerminalSize newSizeTree
+
+        (k, Nothing) -> do
+          writeTVar dConns (ksDelete k conns)
+          newSizeTree <- modifyAndReturnTVar dSizes (deleteMap k)
+          maybeUpdateTerminalSize newSizeTree
+
   where
-    waitForBelt :: KeyedSet Client -> STM (Int, Maybe Belt)
-    waitForBelt ks = asum
+    waitForTake :: KeyedSet Client -> STM (Int, Maybe ClientTake)
+    waitForTake ks = asum
                    $ fmap (\(k,c) -> (k,) <$> Term.take c)
                    $ mapToList
                    $ _ksTable ks
+
+    maybeUpdateTerminalSize :: IntMap TermSize -> STM (Maybe ClientTake)
+    maybeUpdateTerminalSize newSizeTree = do
+      let termSize = foldr minTermSize (TermSize 1024 1024) newSizeTree
+      curSize <- readTVar dMinSize
+      if curSize == termSize
+        then pure Nothing
+        else do
+          writeTVar dMinSize termSize
+          pure $ Just (ClientTakeSize termSize)
+
+    modifyAndReturnTVar :: TVar a -> (a -> a) -> STM a
+    modifyAndReturnTVar var fun = do
```

Comment:
minor nit: consider renaming to modifyAndReadTVar. Also what's the stricntess semantics of this; is it like modifyTVar or modifyTVar'?

Reply:
Changed name, added `'` to name, and made the call `let !post = ...` to force strictness.